src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -637,6 +654,99 @@ public bool IsTypeDesc
 
             return (MethodTable*)m_asTAddr;
         }
+
+        /// <summary>
+        /// Gets a <see cref="MethodTable"/> pointer for the current type handle.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public MethodTable* GetMethodTable()
+        {
+            if (IsTypeDesc)
+            {
+                return AsTypeDesc()->GetMethodTable();
+            }
+
+            return (MethodTable*)m_asTAddr;
+        }
+    }
+
+    internal unsafe struct TypeDesc
+    {
+        /// <summary>
+        /// <para>
+        /// The low-order 8 bits of this flag are used to store the <see cref="CorElementType"/>,
+        /// which discriminates what kind of <see cref="TypeDesc"/> this is.
+        /// </para>
+        /// <para>The remaining bits are available for flags.</para>
+        /// </summary>
+        private readonly uint TypeAndFlags;
+
+        /// <summary>
+        /// Gets a <see cref="MethodTable"/> pointer for the current type desc.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public MethodTable* GetMethodTable()
+        {
+            // The native version here has a check for IsGenericVariable, which causes this method to return null.
+            // This path is never taken by TypedReference, as it always assumes the returned MethodTable* pointer
+            // is not null. If adding code that needs this path to be available as well, it needs to be ported too.
+            //
+            // if (IsGenericVariable)
+            // {
+            //     return null;
+            // }
+
+            if (GetInternalCorElementType() == CorElementType.ELEMENT_TYPE_FNPTR)
+            {
+                return RuntimeTypeHandle.GetElementTypeMethodTable(CorElementType.ELEMENT_TYPE_U);
+            }
+
+            ParamTypeDesc* asParam = (ParamTypeDesc*)Unsafe.AsPointer(ref this);
+
+            if (GetInternalCorElementType() == CorElementType.ELEMENT_TYPE_VALUETYPE)
+            {
+                return asParam->m_Arg.AsMethodTable();
+            }
+
+            return asParam->GetTemplateMethodTableInternal();
```

Comment:
The TemplateMethodTable seems to be dead code. You should be able to delete it here and in the VM. (Delete `m_TemplateMT` and all related code. Delete `OwnsTemplateMethodTable` and fix all callers as if it returned `false`.)

Reply:
Just so I understand, does this mean that the previous `if (GetInternalCorElementType() == CorElementType.ELEMENT_TYPE_VALUETYPE)` check will always pass here? As in, should I also remove it (adding an assert for the cor element type) and just always return `asParam->m_Arg.AsMethodTable()`? 🤔

Reply:
The only typedescs that can legally show up here are pointers and function pointers. It is the same situation as with arrays in previous PR.

It means that you can just check for the Typedesc and return MethodTable for UIntPtr without inspecting the Typedesc further.

> `if (GetInternalCorElementType() == CorElementType.ELEMENT_TYPE_VALUETYPE)`

This case is for handling internal types for interop stubs. These types are internal interop implementation detail and they won't ever show up in managed code. So handling for this case needs to be stay in the unmanaged implementation, but it does not need to be in the managed one.

Reply:
I see, thank you for the additional explanation 🙂
I've simplified the managed path in https://github.com/dotnet/runtime/pull/70055/commits/24976fbf3413367c4da1920648830e2a00873656 and removed the template method table code in a359f578f752ce284639754fffeb1721a2d68a91.